### PR TITLE
New union representation

### PR DESCRIPTION
**Pull request by henridf**
More commits to come:

- [x] handle null union values
- [x] handle containers in unions

(And for completeness, zjson and ndjson (including type "inference") are also outstanding but will  come on a separate PR).

### CODE REVIEW

**Comment by henridf** [comment]
Ready for review

**Comment by henridf** [comment]
rebased and resolved conflicts from #290

File: zng/type.go [hunk 1]
```diff
@@ -216,6 +216,15 @@ func ContainedType(typ Type) (Type, []Column) {
...
 }
 
+func IsUnionType(typ Type) bool {
+	switch typ.(type) {
+	case *TypeUnion:
+		return true
+	default:
+		return false
+	}
+}
```
Comment (nwt) [inline zng/type.go]: I don't think this is worth adding since folks can just say `_, ok := typ.(*zng.TypeUnion)`, but if you do want it, please shorten to
```go
_, ok := typ.(*zng.TypeUnion)
return ok
```

File: zng/type.go [hunk 2]
```diff
@@ -216,6 +216,15 @@ func ContainedType(typ Type) (Type, []Column) {
...
+
 func IsContainerType(typ Type) bool {
 	switch typ.(type) {
 	case *TypeSet, *TypeArray, *TypeRecord:
```
Comment (nwt) [inline zng/type.go]: It's confusing that a union is encoded as a container but `IsContainerType((*TypeUnion)(nil))` is false. Can we make it true?

File: zio/zjsonio/reader.go [hunk 3]
```diff
@@ -165,6 +165,34 @@ func decodeType(columns []interface{}) (string, error) {
...
+	}
+	if utyp, ok := typ.(*zng.TypeUnion); ok {
+		b, typ, index, err = utyp.SplitZng(b)
+		if err != nil {
+			return err
+		}
+		builder.BeginContainer()
+		defer builder.EndContainer()
+		var a [8]byte
+		n := zcode.EncodeCountedVarint(a[:], int64(index))
```
Comment (nwt) [inline zio/zjsonio/reader.go]: Should this be zcode.EncodeCountedUvarint?

File: zio/zngio/parser.go [hunk 4]
```diff
@@ -85,18 +85,31 @@ func zngParseContainer(builder *zcode.Builder, typ zng.Type, b []byte) ([]byte,
...
-
+	if utyp, ok := typ.(*zng.TypeUnion); ok {
+		b, typ, index, err = utyp.SplitZng(b)
+		if err != nil {
+			return nil, err
+		}
+		builder.BeginContainer()
+		defer builder.EndContainer()
+		var a [8]byte
+		n := zcode.EncodeCountedVarint(a[:], int64(index))
```
Comment (nwt) [inline zio/zngio/parser.go]: Should this be zcode.EncodeCountedUvarint?

File: zng/type.go (hunk 1 above)
Comment (henridf) [reply zng/type.go]: This is here because `_, ok := type.(foo)` doesn't compose in boolean expressions. Shortening per your suggestion.

File: zio/zjsonio/reader.go (hunk 3 above)
Comment (henridf) [reply zio/zjsonio/reader.go]: That seems like a good idea, changing.

File: zio/zngio/parser.go (hunk 4 above)
Comment (henridf) [reply zio/zngio/parser.go]: +1

File: zng/type.go (hunk 2 above)
Comment (henridf) [reply zng/type.go]: @nwt good question. 

Since you refer to "union is encoded as a container" I think the premise of your comment is that `zng.IsContainerType()` is a question about the bzng encoding of a type. If that's the case, then I agree that answering "false" to `zng.IsContainerType(union)` is confusing - in fact it's not just confusing, it's incorrect. 

However (and I think more importantly), if `zng.IsContainerType()` is a question about the bzng encoding of the type, then it should presumably be moved elsewhere (bzngio?) as it's not the zng package's responsibility to answer questions about how bzng encoding works.

A counter-argument could be made that `IsContainerType` is a question about the type itself rather than of its bzng encoding (as per the type-level distinction "primitive" vs "container" that we've talked about elsewhere). This interpretation is supported by the use of `IsContainerType()` (via `Value.IsContainer()`) in some zql procs. Under this interpretation answering false to `zng.IsContainerType(union)` is correct. 

I think a clean resolution could be to have both a `zng.IsContainerType()` (that answers the question about the type itself) and a `bzngio.IsContainerEncoded()` (or some better name) that answers the question about the physical layer encoding, and use the right one in the right place. For clarity we might also want to consider using a different terminology for on-the-wire bzng "containers" vs type-level "containers", but that's a broader question.

Long story short: @nwt let me know how I should proceed and I will do it in this PR or a separate one as appropriate.

Comment (henridf) [reply zng/type.go]: Another element for context here: I was just reminded of `zng.ContainedType()`, that seems to line up with the second interpretation above...

Comment (nwt) [reply zng/type.go]: My view is that package zng directly reflects the BZNG spec and that ought to continue unless there's a good reason. Diverging here adds a surprising quirk, and I'm not seeing the good reason.

> A counter-argument could be made that IsContainerType is a question about the type itself rather than of its bzng encoding (as per the type-level distinction "primitive" vs "container" that we've talked about elsewhere). Under this interpretation answering false to zng.IsContainerType(union) is correct.

The flaw I see in this argument is that a union value does in fact contain another value. That's my definition of a container. 😃

Comment (henridf) [reply zng/type.go]: > The flaw I see in this argument is that a union value does in fact contain another value. 

I disagree, but this may be a bikeshed. 

Regardless, how would you like me to proceed? I can make the change you request, but if so please let me know what (if anything) should be done about `zng.ContainedType`. If a union type is a container, then containedType should return the "inner" type of the union type, which can't be done based on the type alone...

File: zng/union.go [hunk 5]
```diff
@@ -44,25 +43,53 @@ func (t *TypeUnion) String() string {
...
+	}
+	if !it.Done() {
+		return nil, -1, nil, ErrBadValue
+	}
+	return inner, int64(index), v, nil
+}
+
+// SplitZng takes a zng encoding of a value of the receiver's type and returns the
+// concrete type of the value, its index into the union type, and the concrete value
+// encoding.
```
Comment (mattnibs) [inline zng/union.go]: /s/the concrete value encoding/the raw value encoding

File: zng/union.go [hunk 6]
```diff
@@ -44,25 +43,53 @@ func (t *TypeUnion) String() string {
 }
 
 func (t *TypeUnion) Parse(in []byte) (zcode.Bytes, error) {
+	panic("TypeUnion.Parse shouldn't be called")
+}
+
+// SplitBzng takes a bzng encoding of a value of the receiver's type and
+// returns the concrete type of the value, its index into the union
+// type, and the concrete value encoding.
```
Comment (mattnibs) [inline zng/union.go]:
```suggestion
// type, and the raw value encoding.
```

Comment (henridf) [reply zng/union.go]: i just dropped "concrete" but didn't add raw as it seems to describe something slightly different...

**Comment by henridf** [comment]
I didn't hear back on `zng.ContainedType()` so going ahead and making @nwt 's requested change to say union is a container, and leaving the ensuing quirk of `zng.ContainedType()`

File: zng/recordval.go [hunk 7]
```diff
@@ -189,6 +196,64 @@ func checkVector(typ *TypeArray, body zcode.Bytes) error {
...
+	case *TypeUnion:
+		if !container {
+			return &RecordTypeError{Name: "<union value>", Type: v.String(), Err: ErrNotContainer}
+		}
+		if err := checkUnion(v, body); err != nil {
+			return err
+		}
+	default:
+		if container {
+			return &RecordTypeError{Name: "<array element>", Type: v.String(), Err: ErrNotPrimitive}
```
Comment (nwt) [inline zng/recordval.go]: s/array/union/

Comment (henridf) [reply zng/recordval.go]: https://github.com/brimsec/zq/pull/309
